File: r/panel_anomaly_detection_parquet.R
#!/usr/bin/env Rscript
# Point-in-time financial statement panel analysis + anomaly detection, same
# as panel_anomaly_detection.R, but reading the SEC `num`/`sub` tables from
# Parquet files instead of querying PostgreSQL.
#
# Sources panel_anomaly_detection.R to reuse its panel-building and
# anomaly-scoring logic (Altman Z''-Score, Beneish M-Score, Benford's Law)
# so the two scripts run an identical analysis over the same data, differing
# only in the data-access layer being timed.
#
# Requires: arrow, dplyr, tidyr, optparse (plus DBI/RPostgres, transitively
# loaded from panel_anomaly_detection.R even though no DB connection is made)
#
# Usage:
#   Rscript panel_anomaly_detection_parquet.R \
#       --num-path ./sec_data/parquet/num --sub-path ./sec_data/parquet/sub \
#       --top-n 25
#
#   Wildcards are also supported:
#   Rscript panel_anomaly_detection_parquet.R \
#       --num-path "./sec_data/parquet/num/year=*/num.parquet" \
#       --sub-path "./sec_data/parquet/sub/*.parquet"

suppressPackageStartupMessages({
  library(arrow)
  library(dplyr)
  library(optparse)
})

get_script_dir <- function() {
  cmd_args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", cmd_args, value = TRUE)
  if (length(file_arg) == 1) return(dirname(normalizePath(sub("^--file=", "", file_arg))))
  getwd()
}

# brings in build_panel, compute_altman_z, compute_beneish_m, compute_benford,
# run_anomaly_detection, INSTANT_TAGS/DURATION_TAGS, etc. Its own main() is
# guarded by is_main() and will not run when sourced from here.
source(file.path(get_script_dir(), "panel_anomaly_detection.R"))

resolve_paths <- function(path_spec) {
  if (!grepl("[*?\\[]", path_spec)) return(path_spec)
  matches <- sort(Sys.glob(path_spec))
  if (length(matches) == 0) stop(sprintf("No files matched pattern: %s", path_spec))
  matches
}

read_parquet_multi <- function(path_spec, columns) {
  paths <- resolve_paths(path_spec)

  if (length(paths) == 1 && dir.exists(paths)) {
    ds <- arrow::open_dataset(paths)
    return(as.data.frame(dplyr::collect(dplyr::select(ds, dplyr::all_of(columns)))))
  }

  frames <- lapply(paths, function(p) {
    t <- arrow::read_parquet(p)
    as.data.frame(t[, columns, drop = FALSE])
  })
  do.call(rbind, frames)
}

parse_cli_args <- function() {
  option_list <- list(
    make_option("--num-path", type = "character",
                help = "Path to num Parquet data: a file, a directory, or a wildcard pattern (e.g. 'sec_data/num/year=*/num.parquet')"),
    make_option("--sub-path", type = "character",
                help = "Path to sub Parquet data: a file, a directory, or a wildcard pattern (e.g. 'sec_data/sub/*.parquet')"),
    make_option("--form", default = "10-K", help = "Filing form type to analyze"),
    make_option("--min-benford-obs", type = "integer", default = 30,
                help = "Minimum reported facts per company to score with Benford's Law"),
    make_option("--top-n", type = "integer", default = 25,
                help = "Number of top flagged filings to print"),
    make_option("--output", default = NULL, help = "Optional CSV path for flagged filings")
  )
  opt <- optparse::parse_args(OptionParser(option_list = option_list),
                               args = commandArgs(trailingOnly = TRUE))
  if (is.null(opt$`num-path`) || is.null(opt$`sub-path`)) {
    stop("--num-path and --sub-path are required")
  }
  opt
}

fetch_data <- function(opt) {
  start <- Sys.time()

  sub <- read_parquet_multi(opt$`sub-path`, columns = c("adsh", "cik", "name", "fy", "fp", "form"))
  sub <- sub[sub$form == opt$form, ]

  num <- read_parquet_multi(opt$`num-path`, columns = c("adsh", "tag", "ddate", "qtrs", "uom", "coreg", "value"))
  num <- num[num$coreg == "", ]

  is_instant <- num$tag %in% INSTANT_TAGS & num$qtrs == 0
  is_duration <- num$tag %in% DURATION_TAGS & num$qtrs == 4
  num <- num[is_instant | is_duration, ]

  df <- merge(num, sub, by = "adsh")[, c("adsh", "cik", "name", "fy", "fp", "tag", "value")]

  elapsed <- as.numeric(difftime(Sys.time(), start, units = "secs"))
  list(df = df, elapsed = elapsed)
}

main <- function() {
  opt <- parse_cli_args()

  fetched <- fetch_data(opt)
  ad <- run_anomaly_detection(fetched$df, opt$`min-benford-obs`)
  print_results(fetched$elapsed, nrow(fetched$df), ad, opt,
                query_label = "Parquet read/query time: ")
}

if (is_main("panel_anomaly_detection_parquet.R")) {
  main()
}

File: python/panel_anomaly_detection_parquet.py
#!/usr/bin/env python3
"""
Point-in-time financial statement panel analysis + anomaly detection, same
as panel_anomaly_detection.py, but reading the SEC `num`/`sub` tables from
Parquet files instead of querying PostgreSQL.

Reuses the panel-building and anomaly-scoring logic (Altman Z''-Score,
Beneish M-Score, Benford's Law) from panel_anomaly_detection.py so the two
scripts run an identical analysis over the same data, differing only in the
data-access layer being timed.

Usage:
  python panel_anomaly_detection_parquet.py \\
      --num-path ./sec_data/num.parquet --sub-path ./sec_data/sub.parquet \\
      --top-n 25
"""
import argparse
import glob
import re
import time

import pandas as pd

from panel_anomaly_detection import DURATION_TAGS, INSTANT_TAGS, run_anomaly_detection

GLOB_CHARS = set("*?[")


def resolve_paths(path_spec):
    """Expand a wildcard pattern (e.g. 'num/year=*/num.parquet') to a list of
    files; otherwise return the path unchanged (a single file or a directory,
    which pandas/pyarrow already knows how to read as-is)."""
    # Unlike a shell or R's Sys.glob(), Python's glob module does not treat
    # backslash as an escape character -- "\=" is matched as a literal
    # backslash followed by "=", which never matches real paths (e.g.
    # Hive-style "year=2020" partition directories have no backslash).
    # Job schedulers/shells commonly backslash-escape "=" defensively, so
    # unescape any "\X" -> "X" before matching, mirroring real glob(3).
    path_spec = re.sub(r"\\(.)", r"\1", path_spec)

    if not any(c in path_spec for c in GLOB_CHARS):
        return path_spec

    matches = sorted(glob.glob(path_spec, recursive=True))
    if not matches:
        raise FileNotFoundError(f"No files matched pattern: {path_spec}")
    return matches


def read_parquet_multi(path_spec, columns):
    paths = resolve_paths(path_spec)
    if isinstance(paths, list):
        return pd.concat(
            (pd.read_parquet(p, columns=columns) for p in paths),
            ignore_index=True,
        )
    return pd.read_parquet(paths, columns=columns)


def parse_args():
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--num-path", required=True,
                    help="Path to num Parquet data: a file, a directory, or a "
                         "wildcard pattern (e.g. 'sec_data/num/year=*/num.parquet')")
    p.add_argument("--sub-path", required=True,
                    help="Path to sub Parquet data: a file, a directory, or a "
                         "wildcard pattern (e.g. 'sec_data/sub/*.parquet')")
    p.add_argument("--form", default="10-K", help="Filing form type to analyze")
    p.add_argument("--min-benford-obs", type=int, default=30,
                    help="Minimum reported facts per company to score with Benford's Law")
    p.add_argument("--top-n", type=int, default=25,
                    help="Number of top flagged filings to print")
    p.add_argument("--output", default=None, help="Optional CSV path for flagged filings")
    return p.parse_args()


def fetch_data(args):
    start = time.perf_counter()

    sub = read_parquet_multi(args.sub_path, columns=["adsh", "cik", "name", "fy", "fp", "form"])
    sub = sub[sub["form"] == args.form]

    num = read_parquet_multi(
        args.num_path,
        columns=["adsh", "tag", "ddate", "qtrs", "uom", "coreg", "value"],
    )
    num = num[num["coreg"] == ""]

    is_instant = num["tag"].isin(INSTANT_TAGS) & (num["qtrs"] == 0)
    is_duration = num["tag"].isin(DURATION_TAGS) & (num["qtrs"] == 4)
    num = num[is_instant | is_duration]

    df = num.merge(sub, on="adsh", how="inner")[
        ["adsh", "cik", "name", "fy", "fp", "tag", "value"]
    ]

    elapsed = time.perf_counter() - start
    return df, elapsed


def main():
    args = parse_args()

    df, query_time = fetch_data(args)
    print(f"Parquet read/query time:  {query_time:8.2f}s ({len(df):,} facts loaded)")

    result, analysis_time = run_anomaly_detection(df, args.min_benford_obs)
    print(f"Anomaly detection time:    {analysis_time:8.2f}s ({len(result):,} filings scored)")
    print(f"Total time:                {query_time + analysis_time:8.2f}s")

    flagged = result[result["anomaly_count"] > 0].sort_values(
        ["anomaly_count", "m_score"], ascending=[False, False]
    )
    print(f"\n{len(flagged):,} filings flagged by at least one signal\n")

    display_cols = ["adsh", "cik", "name", "fy", "fp", "z_score", "m_score", "benford_chi2", "anomaly_count"]
    print(flagged[display_cols].head(args.top_n).to_string(index=False))

    if args.output:
        flagged.to_csv(args.output, index=False)
        print(f"\nFlagged filings written to {args.output}")


if __name__ == "__main__":
    main()
